packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -421,6 +421,21 @@ function resolve(
     if (shouldConstruct(Component)) {
       inst = new Component(element.props, publicContext, updater);
     } else {
+      if (__DEV__) {
+        if (
+          Component.prototype &&
+          typeof Component.prototype.render === 'function'
+        ) {
```

Comment:
We should make sure we don't trigger this warning for the same component more than once.
If the client-side code doesn't deduplicate them, we also need to add deduplication there too.
Something like

```js
const didWarnAboutBadClass = {};

// ...

if (!didWarnAboutBadClass[componentName]) {
  warning(...)
  didWarnAboutBadClass[componentName] = true
}